### PR TITLE
WIP isort

### PR DESCRIPTION
**Pull request by gforcada**
**Do not merge**

I had some time this morning and worked on it a bit, it's totally unfinished, untested, unchecked, so keep your kittens away from it! :)

### CODE REVIEW

**Comment by saily** [comment]
Please have a look to the branch again which is almost complete now including tests. 
But we have several problems with `isort` which looks like bugs on `isort` to me. 

1. `from foo.bar import this_is_a_very_long_baz_which_for_sure_exceeds_line_length` is not wrapped on `--force_single_line_imports` and if you omit it it's also incorrect, see below:
    ```
@@ -1,7 +1,7 @@
-from foo.bar import baz
-from foo.bar import \
+    baz,
     this_is_a_very_long_baz_which_for_sure_exceeds_line_length
-from foo.car import \
-    this_is_a_very_long_baz_which_for_sure_exceeds_line_length
+)
+from foo.bar import (
+from foo.car import this_is_a_very_long_baz_which_for_sure_exceeds_line_length

-import baz+import baz
```

2. `isort` requires an extra line which is wrong IMHO: 
    ```
plone/recipe/codeanalysis/tests/test_zptlint.py:after	2015-09-03 22:45:52.945262
@@ -1,13 +1,13 @@
 # -*- coding: utf-8 -*-
 from plone.recipe.codeanalysis.testing import CodeAnalysisTestCase
-from plone.recipe.codeanalysis.zptlint import console_script
-from plone.recipe.codeanalysis.zptlint import ZPTLint
+from plone.recipe.codeanalysis.zptlint import ZPTLint, console_script
 from shutil import rmtree
 from tempfile import mkdtemp
 from testfixtures import OutputCapture

 import os
 import unittest
+   <-- this is wrong!


 # EXTRAS_INSTALLED is an environment variable that we set on
```

**Comment by gforcada** [comment]
@saily then file issues/PR on isort :)

As for wrapping long imports I lately prefer to just append a ``# noqa`` on them. They are just imports and these days all IDE hide them by default anyway.

File: README.rst [hunk 1]
```diff
@@ -435,6 +426,25 @@ The recipe supports the following options:
...
+    Allows you to change the way imports should be sorted. Default is to follow
+    `plone.api conventions`_. See `isort documentation for available options
+    <http://pypi.python.org/pypi/isort>`_::
+
+        ...
+        isort = True
+        isort-force-alphabetical-sort = True
+        isort-force_single_line_imports = True
+        isort-diff = True
+
```
Comment (gforcada) [inline README.rst]: @saily what's the benefit of this compared to just passing the command line options in a single variable as my original code was doing?

That's just so much verbose and on top of that you can not play with isort on the commandline, check if you like the result and then **verbatim** copy that to your buildout.cfg and you are good to go.

With this approach you have to add the ``isort-`` prefix, the `` = True`` and so on...

Comment (saily) [inline README.rst]: i just wanted to keep it consistent inside our recipe. we're using this also for flake8.

**Comment by do3cc** [comment]
Why do we offer the option `isort-force_single_line_imports` at all? It is not plone.api conform to do it differently

**Comment by saily** [comment]
@do3cc: good point. Can we define an allowed subset of options?

@gforcada: `# noqa` should be appended automatically if it breaks the `line-length` option

**Comment by gforcada** [comment]
@do3cc well I guess that if we want to be useful outside Plone it should be configurable, not everyone agrees on plon.api coding style... I see no problem on making it a default option, as long as it can be overriden, at least z3c.dependencychecker will welcome it

**Comment by gforcada** [comment]
@saily then another issue to file on isort :) (about the noqa)

**Comment by saily** [comment]
what i see here, it's just too early and not that easy to integrate `isort` into this recipe.

**Comment by do3cc** [comment]
https://github.com/timothycrosley/isort/pull/326

**Comment by do3cc** [comment]
I'd prefer to only offer configuration options that are plone.api valid, together with the option to pass on arbitrary number of options with another setting.
This way, people can still break plone.api.
Every option we support explicitly must be maintained. When upstream decides to rename the option, we have to invest time into migration, for a use case we do not use ourselves.

**Comment by do3cc** [comment]
My other comment was a bit too sparse.
I added the option to not break imports into multilines, but to add a # NOQA at the end with this pr
https://github.com/timothycrosley/isort/pull/326

**Comment by do3cc** [comment]
My spare time ran out, here is the bug report for the newlines in isort 
https://github.com/timothycrosley/isort/pull/327

**Comment by gforcada** [comment]
On the meantime I released https://pypi.python.org/pypi/flake8-isort if you want to give it a try...

At the end I find way more simpler and easier to just write flake8 plugins...

**Comment by gforcada** [comment]
I'm closing this one as IMHO the flake8 plugin is the way to go (and of course fixing the problems upstream on isort)

**Pull request closed without merging**